src/Datasource/QueryTrait.php

```diff
@@ -319,13 +320,13 @@ public function toArray(): array
      * If the third argument is set to true, it will erase previous map reducers
      * and replace it with the arguments passed.
      *
-     * @param callable|null $mapper The mapper callable.
-     * @param callable|null $reducer The reducing function.
+     * @param \Closure|null $mapper The mapper callable.
+     * @param \Closure|null $reducer The reducing function.
      * @param bool $overwrite Set to true to overwrite existing map + reduce functions.
      * @return $this
      * @see \Cake\Collection\Iterator\MapReduce for details on how to use emit data to the map reducer.
      */
-    public function mapReduce(?callable $mapper = null, ?callable $reducer = null, bool $overwrite = false)
```

Comment:
This one is not ok, for instance.

Reply:
If we want to reap the benefits of making it faster we could remove the internal type checks for callable and only keep them at boundary methods such as this one. 
The boundary methods can convert to closure to make the internals faster